### DISCUSSION TITLE
Empowering Characters: Dynamic Adaptation, Autonomous Growth, and Enhanced Alignment in Long Conversations

### DISCUSSION THREAD

**Discussion by skatardude10** [oobabooga/text-generation-webui#2811] [Ideas]
Good evening/morning!

The idea is to enable LLMs/characters to dynamically modify variables in their base prompt/profile, enabling better alignment with decisions and changes throughout long conversatios that may conflict with the base prompt, and also allow characters to evolve and grow with autonomy and greater depth. 

Certain models perform very well in following long-term stories during extended chats with superbooga, but they still fight or resist major base prompt conflicting changes/decisions that fall out of context after many messages. Manually editing the changes into the base prompt after significant decisions or changes mid conversation works fantastically though, enabling characters to adapt, ensuring they remain consistent and can evolve throughout the story/chat- but it's manual and I know it's possible to let it define its own major changes. 

I tried one approach, which was prompting the LLM to indicate what changes it would like to make to it's base prompt, but it failed miserably.

To implement this in a way that might work though, I recall seeing code on GitHub or Huggingface (though I can't find it now) that shows the layers and weights of different words when you enter a prompt with probabilities for various next tokens. This code could potentially be adapted to help an LLM/character evolve itself in real-time by applying those probabilities to variables in the character's base prompt/character card.

Here's how it could work: When building a character card or base prompt, users could assign certain tokens, words, or entire phrases as variables that define the character, situation, or anything in the base prompt that could shift. These variables could change based on either user-defined values, or predictions made by the LLM based on token probabilities.  

Ex: [Personality: Fun] is probably static so probably not a variable, but [Friends with X] can easily be variable -- Character is { ... } With Character B... Predict probability of { ... } In context by feeding full context into LLM, with ?snippet of base context surrounding the variable to get highest probability for what to change the variable within the current full context. Variables could even be assigned as top 1-5 probabilities... (Ex:  1x {Hates}   or  2x {Hates/Dislikes}  or  feels  3x {neutral,friendly,unsure} to give the LLM the autonomy to provide even more depth to itself and major changes based on the context of the conversation. 

The simplest trigger would probably just be either manual button to click to do automatic prediction of alternatives to all variables when clicked, or an automatic variable evaluation trigger after a certain number of messages could be implemented. Or, a semantic match or dissimilarity detection between a variables current assigned values and highest probability prediction of the token that fills the a blank variable in short or full context. 

Overall, the idea is to have the LLM/character adapt and evolve based on the context of the conversation. Manually editing the base prompt mid conversation does this effectively. By enabling the user to define variables in a base prompt, and empowering the LLM itself to make determinations as to what these variables should be defined as in context I think would make for much more dynamic characters that track more effectively versus LLMs that have strayed from their base prompt- Allowing them to better align with their decisions and grow throughout the story/chat automatically, which would be a lot of fun and potentially very interesting.

**Comment by tensiondriven** [comment]
I think the folks over at KoboldAI have given this quite a bit of thought - you may want to check in on the kobold subreddit.
